libraries/c_sdk/standard/mqtt/src/iot_mqtt_subscription.c

```diff
@@ -318,70 +299,75 @@ IotMqttError_t _IotMqtt_AddSubscriptions( _mqttConnection_t * pMqttConnection,
+    contextIndex = _IotMqtt_getContextIndexFromConnection( pMqttConnection );
 
-    for( i = 0; i < subscriptionCount; i++ )
+    if( IotMutex_Take( &( connToContext[ contextIndex ].subscriptionMutex ) ) == true )
     {
-        /* Check if this topic filter is already registered. */
-        topicMatchParams.pTopicName = pSubscriptionList[ i ].pTopicFilter;
-        topicMatchParams.topicNameLength = pSubscriptionList[ i ].topicFilterLength;
-        pSubscriptionLink = IotListDouble_FindFirstMatch( &( pMqttConnection->subscriptionList ),
-                                                          NULL,
-                                                          _topicMatch,
-                                                          &topicMatchParams );
-
-        if( pSubscriptionLink != NULL )
-        {
-            pNewSubscription = IotLink_Container( _mqttSubscription_t, pSubscriptionLink, link );
-
-            /* The lengths of exactly matching topic filters must match. */
-            IotMqtt_Assert( pNewSubscription->topicFilterLength == pSubscriptionList[ i ].topicFilterLength );
-
-            /* Replace the callback and packet info with the new parameters. */
-            pNewSubscription->callback = pSubscriptionList[ i ].callback;
-            pNewSubscription->packetInfo.identifier = subscribePacketIdentifier;
-            pNewSubscription->packetInfo.order = i;
-        }
-        else
+        for( i = 0; i < subscriptionCount; i++ )
         {
-            /* Allocate memory for a new subscription. */
-            pNewSubscription = IotMqtt_MallocSubscription( sizeof( _mqttSubscription_t ) +
-                                                           pSubscriptionList[ i ].topicFilterLength );
+            /* Check if this topic filter is already registered. */
+            topicMatchParams.pTopicName = pSubscriptionList[ i ].pTopicFilter;
+            topicMatchParams.topicNameLength = pSubscriptionList[ i ].topicFilterLength;
+            /* Finding the matching subscription if it exists. */
+            matchedIndex = IotMqtt_FindFirstMatch( &( connToContext[ contextIndex ].subscriptionArray ), 0, &topicMatchParams );
 
-            if( pNewSubscription == NULL )
+            if( matchedIndex != -1 )
             {
-                status = IOT_MQTT_NO_MEMORY;
-                break;
+                IotMqtt_Assert( connToContext[ contextIndex ].subscriptionArray[ matchedIndex ].topicFilterLength == pSubscriptionList[ i ].topicFilterLength );
+
+                /* Replace the callback and packet info with the new parameters. */
+                connToContext[ contextIndex ].subscriptionArray[ matchedIndex ].callback = pSubscriptionList[ i ].callback;
+                connToContext[ contextIndex ].subscriptionArray[ matchedIndex ].packetInfo.identifier = subscribePacketIdentifier;
+                connToContext[ contextIndex ].subscriptionArray[ matchedIndex ].packetInfo.order = i;
```

Comment:
what is this order used for?

Reply:
order is used to maintain the order ,when multiple subscriptions are there in SUBSCRIBE packet.

Reply:
ok. Got it